### PR TITLE
Add Unwrap derive

### PR DESCRIPTION
**Pull request by Monadic-Cat**
```rust
#[derive(::derive_more::Unwrap)]
enum Lol {
    Haha(i32),
    Hm,
}

fn main() {
    let a = Lol::Haha(10);
    println!("Results: {:?}", a.unwrap_haha())
}
```
prints `Results: 10`, and switching to `a.unwrap_hm()` prints this message:
> thread 'main' panicked at 'called `Lol::unwrap_hm()` on a `Haha` value', src/main.rs:18:33
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

### CODE REVIEW

**Comment by raffomania** [comment]
This is very nice. Please let me know if I can help :)

**Comment by JelteF** [comment]
I think this feature sounds nice, but it needs some tests.

**Comment by Monadic-Cat** [comment]
I can add docs for this some time in the next week or so, depending on when I'm free, and maybe I can do tests based on whatever `IsVariant` has? This is pretty much a copy and paste of that macro with a few tweaks.

**Comment by Monadic-Cat** [comment]
While writing the docs based on `IsVariant`'s docs, I noticed and tested `#[unwrap(ignore)]`, which doesn't work. Its use of `.enabled_variant_data()` for some reason unconditionally reports all variants. It works with `IsVariant`, though, so I'm trying to pin down what's different between the two.

File: src/unwrap.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use syn::{DeriveInput, Fields, Ident, Result};
+
+pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStream> {
+    let state = State::with_attr_params(
+        input,
+        trait_name,
+        quote!(),
+        String::from("unwrap_variant"),
```
Comment (JelteF) [inline src/unwrap.rs:12]: Should be `unwrap` to fix `#[unwrap(ignore)]`.

Comment (Monadic-Cat) [reply src/unwrap.rs]: That works, thank you!

**Comment by Monadic-Cat** [comment]
I've now added documentation and tests based on what's there for the `IsVariant` derive, and fixed a, in retrospect, rather obvious bug.

**Comment by Monadic-Cat** [comment]
Looking over CI failures now:

Do you have an MSRV of 1.36 (going based on the CI failure)?
If so, this feature can't use the `#[track_caller]` attribute in its generated code, which will make the panic messages less helpful.

Lemme `rustfmt` the new stuff right now...

**Comment by JelteF** [comment]
> Do you have an MSRV of 1.36 (going based on the CI failure)?

Yes, so far that's the MSRV indeed. What version is the first one that supports `track_caller`?

I think you can probably do something similar to what we do for nightly detection: 
https://github.com/JelteF/derive_more/blob/d335514f9f732f2e1b45c4bc7f3af20502a87c5d/build.rs#L26-L35

i.e. create a feature called `track_caller` (or something) when the rust version supports `track_caller`. Then only add the `track_caller` attribute when the `track_caller` feature is enabled.

**Comment by Monadic-Cat** [comment]
> Yes, so far that's the MSRV indeed. What version is the first one that supports `track_caller`?

That would be Rust [1.46](https://blog.rust-lang.org/2020/08/27/Rust-1.46.0.html#track_caller).

> I think you can probably do something similar to what we do for nightly detection:

Done!

**Comment by JelteF** [comment]
The doctests are failing. To test them locally you need to use my fork of `cargo external-doc`. Which you can install like:
```
cargo install --git https://github.com/JelteF/cargo-external-doc --force
```

**Comment by Monadic-Cat** [comment]
That looks like everything!

**Review by JelteF — APPROVED** [review]

**Comment by JelteF** [comment]
This is included in v0.99.16

**Pull request merged**